### PR TITLE
Download data button added to manage sensor admin pane

### PR DESCRIPTION
**Pull request by akampen**

### CODE REVIEW

File: src/components/Admin/UserPage/DownloadCSV.tsx [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+  /**
+   * Drop-down menu for months of the year
+   * @param value - state variable this input displays and changes
+   * @param setValue - function that sets the value
+   */
+  const MonthInput: ({value, setValue}: InputProps) => JSX.Element = ({
+    value,
+    setValue,
+  }: InputProps) => {
+    const labels = [
```
Comment (acgrobman) [inline src/components/Admin/UserPage/DownloadCSV.tsx:281]: Will this translate?

File: src/components/Admin/UserPage/DownloadCSV.tsx [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+ * in the administrative pane.
+ */
+const DownloadCSVModal: () => JSX.Element = () => {
+  const {t} = useTranslation('administration');
+  /* --------------- State maintenance variables ------------------------ */
+  const {isOpen, onOpen, onClose} = useDisclosure();
+  const [startYear, setStartYear] = useState(0);
+  const [startMonth, setStartMonth] = useState(0);
+  const [startDay, setStartDay] = useState(0);
+  const [endYear, setEndYear] = useState(0); // Change to timestamp
```
Comment (acgrobman) [inline src/components/Admin/UserPage/DownloadCSV.tsx:249]: Is this comment referring to an issue?

File: src/components/Admin/UserPage/DownloadCSV.tsx [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+  const [body, setBody] = useState<BodyElement[]>([]);
+  const [header, setHeader] = useState<HeaderElement[]>([]);
+  const [progress, setProgress] = useState(0);
+  const [totalSensors, setTotalSensors] = useState(1);
+  const [filename, setFilename] = useState('');
+  const [fetchingData, setFetchingData] = useState(false);
+  const [readyForDownload, setReadyForDownload] = useState(false);
+  /* ------------------------------------------------------------------- */
+
+  function fetchData() {
```
Comment (CarsonHerness) [inline src/components/Admin/UserPage/DownloadCSV.tsx:106]: Add tsdoc

File: src/components/Admin/UserPage/DownloadCSV.tsx [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+      {label: 'Longitude', key: 'longitude'},
+    ];
+
+    /**
+     * This function actually fetches the data from firestore and populates the
+     * list that represents the body of the CSV file
+     * @param body - list of body elements to populate with the data
+     * @returns the updated body with the elements from Firestore
+     */
+    async function getData(body: BodyElement[]) {
```
Comment (CarsonHerness) [inline src/components/Admin/UserPage/DownloadCSV.tsx:135]: Can this function be declared outside of the `fetchData` function?

File: src/components/Admin/UserPage/DownloadCSV.tsx [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+    const endDate = new Date(endYear, endMonth - 1, endDay);
+    const endDateString = endDate.toISOString();
+
+    setFilename('pm25_' + startDateString + '_to_' + endDateString + '.csv');
+
+    const newBody: BodyElement[] = [];
+    const newHeaders: HeaderElement[] = [
+      {label: 'Timestamp', key: 'timestamp'},
+      {label: 'Name', key: 'name'},
+      {label: 'PM 2.5', key: 'pm25'},
```
Comment (CarsonHerness) [inline src/components/Admin/UserPage/DownloadCSV.tsx:122]: nit: remove space between `PM` and `2.5`

File: src/components/Admin/UserPage/DownloadCSV.tsx [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+interface InputProps {
+  value: number;
+  setValue: React.Dispatch<React.SetStateAction<number>>;
+}
+
+/**
+ * Component for the button that can be clicked to download sensor data. Used
+ * in `DownloadCSVModal` component which appear son the manage sensor page for admins.
+ */
+const DownloadCSVButton: ({
```
Comment (CarsonHerness) [inline src/components/Admin/UserPage/DownloadCSV.tsx:78]: Move this to a separate file

File: src/components/Admin/UserPage/DownloadCSV.tsx [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+    clearFields();
+    onClose();
+  }
+
+  /**
+   * Drop-down menu for months of the year
+   * @param value - state variable this input displays and changes
+   * @param setValue - function that sets the value
+   */
+  const MonthInput: ({value, setValue}: InputProps) => JSX.Element = ({
```
Comment (CarsonHerness) [inline src/components/Admin/UserPage/DownloadCSV.tsx:277]: Move the `MonthInput` and `DayInput` into a separate file

File: src/components/Admin/UserPage/DownloadCSV.tsx [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+        </Select>
+      </Box>
+    );
+  };
+
+  /**
+   * Check that input dates are valid, sets error accordingly
+   */
+  useEffect(() => {
+    const startDate = new Date(startYear, startMonth - 1, startDay);
```
Comment (CarsonHerness) [inline src/components/Admin/UserPage/DownloadCSV.tsx:358]: Add comment for why `startMonth - 1` is necessary

File: src/components/Admin/UserPage/DownloadCSV.tsx [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+    const endDate = new Date(endYear, endMonth - 1, endDay);
+    if (startDate.getMonth() !== startMonth - 1) {
```
Comment (CarsonHerness) [inline src/components/Admin/UserPage/DownloadCSV.tsx:360]: Add comment explaining what kind of valid dates are checked here (i.e. you can't have a February 30th)

File: src/components/Admin/UserPage/DownloadCSV.tsx [hunk 10]
```diff
@@ -0,0 +1,445 @@
...
+  const [startMonth, setStartMonth] = useState(0);
+  const [startDay, setStartDay] = useState(0);
+  const [endYear, setEndYear] = useState(0); // Change to timestamp
+  const [endMonth, setEndMonth] = useState(0); // Change to timestamp
+  const [endDay, setEndDay] = useState(0);
+  const [error, setError] = useState('');
+  /* ------------------------------------------------------------------- */
+
+  // Reset to starting state
+  function clearFields() {
```
Comment (CarsonHerness) [inline src/components/Admin/UserPage/DownloadCSV.tsx:256]: use tsdoc

File: src/components/Admin/UserPage/DownloadCSV.tsx [hunk 11]
```diff
@@ -0,0 +1,445 @@
...
+    setStartMonth(0);
+    setStartDay(0);
+    setEndYear(0);
+    setEndMonth(0);
+    setEndDay(0);
+    setError('');
+  }
+
+  // Reset starting state and close modal
+  function handleClose() {
```
Comment (CarsonHerness) [inline src/components/Admin/UserPage/DownloadCSV.tsx:267]: use tsdoc

**Comment by CarsonHerness** [comment]
My downloaded document seems to make every entry a string by putting quotes around each entry, I don't think this is intended behavior?

**Comment by CarsonHerness** [comment]
Also add issue to allow users to control which sensors they get data for

File: src/components/Admin/UserPage/DownloadCSV.tsx (hunk 2 above)
Comment (akampen) [reply src/components/Admin/UserPage/DownloadCSV.tsx:249]: Nope, left over from my development

**Comment by akampen** [comment]
> My downloaded document seems to make every entry a string by putting quotes around each entry, I don't think this is intended behavior?

Unfortunately changing the enclosing character is not supported by the library. There was an issue in 2018 linked [here](https://github.com/react-csv/react-csv/issues/68). A dev thought they implemented the ability to change the enclosing character but it doesn't work (as other users have pointed out).

It displays correctly in Numbers and Excel. It seems like something that can be easily addressed on Lotenna and Hillary's side?

**Comment by akampen** [comment]
> Also add issue to allow users to control which sensors they get data for

#536

**Comment by acgrobman** [comment]
> It displays correctly in Numbers and Excel. It seems like something that can be easily addressed on Lotenna and Hillary's side?

@hillaryrguez or @lnwobbi: do y’all know/would y’all be able to test whether the R works with this format?

**Review by CarsonHerness — APPROVED** [review]
This looks great!

**Review by acgrobman — APPROVED** [review]

**Pull request merged**